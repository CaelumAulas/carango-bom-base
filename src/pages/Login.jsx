import React from "react";
import LoginForm from "../components/LoginForm";
import LoginService from "../services/LoginService";

function Login() {
  const onSubmit = async user => {
    try {
      await LoginService.login(user);
    } catch (error) {
      console.error(error);
    }
  };

  return <LoginForm onSubmit={onSubmit} />;
}

export default Login;
