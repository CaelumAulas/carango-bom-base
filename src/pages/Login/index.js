import { useState } from 'react'
import { StyledButton, StyledTextField, useStyles } from './styles'
import useErrors from '../../hooks/useErrors'
import LoginService from '../../services/LoginService'
import { useAuth } from '../../hooks/useAuth'
import { useHistory } from 'react-router-dom'
import { CircularProgress } from '@material-ui/core'
import { validateEmail } from '../../utils/validateEmail'
import { delayFunc } from '../../utils/delayFunc'

export default function Login() {

  const [user, setUser] = useState('')
  const [password, setPassword] = useState('')
  const [error, setError] = useState(false)
  const [loading, setLoading] = useState(false)
  const { saveToken } = useAuth()
  const history = useHistory()

  function isInputFilled(inputValue) {
    if (!!inputValue) {
      return { valid: true }
    }
    return { valid: false, text: "Preencha este campo" }
  }

  const validations = {
    user: inputValue => {
      return { valid: validateEmail(inputValue), text: 'Insira um email válido' }
    },
    password: inputValue => isInputFilled(inputValue)
  }

  const [errors, validateFields, allFieldsValid] = useErrors(validations)

  function atualizaErro(error) {
    setError(error)
  }

  function submissaoValida() {
    const hasEmptyInput = !password || !user;
    atualizaErro(hasEmptyInput)

    if (hasEmptyInput) return

    return allFieldsValid()
  }

  async function login(e) {
    e.preventDefault()

    if (submissaoValida()) {
      setLoading(true)
      const { token } = await LoginService.login({ email: user, password })

      saveToken(token)
      await delayFunc(() => setLoading(false))

      if (token) {
        history.push('/veiculos')
      }
    }
  }

  const classes = useStyles()

  return (
    <div className={classes.root}>
      <form data-testid="form" id="form" onSubmit={login}>
        <h1>Login</h1>
        {!loading ? <>
          <StyledTextField
            placeholder="Digite seu email"
            id="user"
            label="Usuário"
            name="user"
            onChange={({ target: { value } }) => setUser(value)}
            onBlur={validateFields}
            fullWidth
            variant="outlined"
            helperText={errors.user.text}
            error={!errors.user.valid}
          />

          <StyledTextField
            placeholder="Digite sua password"
            id="password"
            type="password"
            label="Senha"
            name="password"
            onChange={({ target: { value } }) => setPassword(value)}
            onBlur={validateFields}
            fullWidth
            variant="outlined"
            helperText={errors.password.text}
            error={!errors.password.valid}
          />

          {error && <p className={classes.msgError}>Preencha todos os campos</p>}

          <StyledButton form="form" fullWidth type="submit" variant="contained">
            Entrar
          </StyledButton>
        </> : <CircularProgress thickness={2} />}
      </form>
    </div>
  )
}