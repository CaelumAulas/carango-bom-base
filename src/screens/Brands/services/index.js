const BrandService = {
  async register(brand) {
    return fetch('https://carango-bom-api.herokuapp.com/marcas', {
      method: 'POST',
      body: JSON.stringify(brand),
    }).then((r) => r.json());
  },

  async update(brand) {
    return fetch(`https://carango-bom-api.herokuapp.com/marcas/${brand.id}`, {
      method: 'PUT',
      body: JSON.stringify(brand),
    }).then((r) => r.json());
  },

  async get(id) {
    return fetch(`https://carango-bom-api.herokuapp.com/marcas/${id}`).then((r) => r.json());
  },

  async getAll() {
    return fetch('https://carango-bom-api.herokuapp.com/marcas').then((r) => r.json());
  },

  async remove(brand) {
    return fetch(`https://carango-bom-api.herokuapp.com/marcas/${brand.id}`, {
      method: 'DELETE',
    }).then((r) => r.json());
  },
};

export default BrandService;
