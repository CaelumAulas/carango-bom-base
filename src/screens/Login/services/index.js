export const login = (username, password) => {
  // TODO: conect with backend API
  console.log(`username, password`, username, password);
};
