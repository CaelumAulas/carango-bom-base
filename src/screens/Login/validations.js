const validations = {
  username: (data) => {
    if (data && data.length >= 4) {
      return { valid: true, showError: false };
    }
    return { valid: false, text: 'UsuÃ¡rio deve ter ao menos 4 caracteres.', showError: true };
  },
  password: (data) => {
    if (data && data.length >= 6) {
      return { valid: true, showError: false };
    }
    return { valid: false, text: 'Senha deve ter ao menos 6 caracteres.', showError: true };
  },
};

export default validations;
