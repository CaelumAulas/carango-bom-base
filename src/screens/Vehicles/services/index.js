const VehicleService = {
  async getAll() {
    return fetch('https://carango-bom-api.herokuapp.com/veiculos').then((r) => r.json());
  },
};

export default VehicleService;
