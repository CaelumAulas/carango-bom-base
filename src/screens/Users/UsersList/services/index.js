import { API_URL } from '../../../../services/constants';
import buildAuthHeader from '../../../../services/buildAuthHeader';
import fetchResponseHandler from '../../../../services/fetchResponseHandler';

export const getAllUsers = async (userJWT) => {
  const requestOptions = {
    method: 'GET',
    headers: buildAuthHeader(userJWT),
  };
  return fetch(`${API_URL}/usuario`, requestOptions).then(fetchResponseHandler);
};
