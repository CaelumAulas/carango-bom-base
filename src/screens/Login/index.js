import React, { useState } from 'react';
import { TextField, Button, Grid } from '@material-ui/core';

import useErrors from '../../hooks/useErrors';

import { login } from './services';
import { useStyles } from './styles';
import validations from './validations';

const Login = () => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const classes = useStyles();
  const [errors, validateFields, isFormValid] = useErrors(validations);

  const onSubmit = (event) => {
    event.preventDefault();

    login(username, password);
  };

  const handleUpdateUsername = (event) => {
    setUsername(event.target.value);
  };

  const handleUpdatePassword = (event) => {
    setPassword(event.target.value);
  };

  return (
    <form onSubmit={onSubmit}>
      <Grid container direction="column" spacing={2} alignContent="center">
        <Grid container item xs={6}>
          <TextField
            fullWidth
            id="username"
            label="UsuÃ¡rio"
            type="text"
            name="username"
            role="textbox"
            onChange={handleUpdateUsername}
            variant="outlined"
            onBlur={validateFields}
            helperText={errors.username.text}
            error={errors.username.showError}
          />
        </Grid>
        <Grid container item xs={6}>
          <TextField
            fullWidth
            id="password"
            label="Senha"
            type="password"
            name="password"
            role="textbox"
            onChange={handleUpdatePassword}
            variant="outlined"
            onBlur={validateFields}
            helperText={errors.password.text}
            error={errors.password.showError}
          />
        </Grid>
        <Grid item xs={6}>
          <Button
            type="submit"
            variant="contained"
            fullWidth
            color="primary"
            className={classes.sumbitButton}
            disabled={!isFormValid()}
          >
            Logar
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};

export default Login;
