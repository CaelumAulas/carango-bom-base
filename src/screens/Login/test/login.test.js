import { render, screen, fireEvent, waitFor, getByText } from '@testing-library/react';

import Login from '..';
import { login } from '../services';

jest.mock('../services');

describe('Login screen', () => {
  it('should match the snapshot', () => {
    const { container } = render(<Login />);

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('Login form', () => {
  let usernameInput;
  let passwordInput;
  let submitButton;

  beforeEach(() => {
    render(<Login />);

    usernameInput = screen.getByRole('textbox', { name: /Usuário/i });
    passwordInput = screen.getByLabelText(/Senha/i);
    submitButton = screen.getByRole('button', { name: /Logar/i });
  });

  it('should render the login form', () => {
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  it('should send the data correctly when a user presses submit', () => {
    fireEvent.change(usernameInput, {
      target: {
        value: 'johndoe',
      },
    });

    fireEvent.focusOut(usernameInput);

    fireEvent.change(passwordInput, {
      target: {
        value: '123456',
      },
    });

    fireEvent.focusOut(passwordInput);

    expect(submitButton).not.toBeDisabled();

    fireEvent.click(submitButton);

    expect(login).toHaveBeenCalledWith('johndoe', '123456');
  });

  it('should only enable the submit button when all fields are filled', () => {
    expect(submitButton).toHaveAttribute('disabled');

    fireEvent.change(usernameInput, {
      target: {
        value: 'johndoe',
      },
    });

    fireEvent.focusOut(usernameInput);

    expect(submitButton).toHaveAttribute('disabled');

    fireEvent.change(passwordInput, {
      target: {
        value: '123456789',
      },
    });

    fireEvent.focusOut(passwordInput);

    expect(submitButton).not.toBeDisabled();
  });

  it('should keep the button disabled when the username don`t meet the requirements', () => {
    fireEvent.change(usernameInput, {
      target: {
        value: 'joe',
      },
    });

    fireEvent.focusOut(usernameInput);

    const errorText = screen.getByText('Usuário deve ter ao menos 4 caracteres.');

    expect(errorText).toBeInTheDocument();

    expect(submitButton).toHaveAttribute('disabled');
  });

  it('should keep the button disabled when the password don`t meet the requirements', () => {
    fireEvent.change(passwordInput, {
      target: {
        value: '123',
      },
    });

    fireEvent.focusOut(passwordInput);

    const errorText = screen.getByText('Senha deve ter ao menos 6 caracteres.');

    expect(errorText).toBeInTheDocument();

    expect(submitButton).toHaveAttribute('disabled');
  });
});
