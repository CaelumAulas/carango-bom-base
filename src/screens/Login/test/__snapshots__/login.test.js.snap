// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen should match the snapshot 1`] = `
<form>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-content-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        role="textbox"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="username"
          id="username-label"
        >
          Usuário
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="username"
            name="username"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                Usuário
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        role="textbox"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Senha
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                Senha
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-sumbitButton-1 MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Logar
        </span>
      </button>
    </div>
  </div>
</form>
`;
